git/types.py

```diff
@@ -4,114 +4,248 @@
-    Union,
-    Any,
-    Optional,
-    Callable,
     TYPE_CHECKING,
     TypeVar,
+    Union,
 )
 
 if sys.version_info >= (3, 8):
     from typing import (  # noqa: F401
         Literal,
-        TypedDict,
         Protocol,
         SupportsIndex as SupportsIndex,
+        TypedDict,
         runtime_checkable,
     )
 else:
     from typing_extensions import (  # noqa: F401
         Literal,
+        Protocol,
         SupportsIndex as SupportsIndex,
         TypedDict,
-        Protocol,
         runtime_checkable,
     )
 
-# if sys.version_info >= (3, 10):
-#     from typing import TypeGuard  # noqa: F401
-# else:
-#     from typing_extensions import TypeGuard  # noqa: F401
-
-PathLike = Union[str, "os.PathLike[str]"]
-
 if TYPE_CHECKING:
-    from git.repo import Repo
     from git.objects import Commit, Tree, TagObject, Blob
+    from git.repo import Repo
 
-    # from git.refs import SymbolicReference
+PathLike = Union[str, "os.PathLike[str]"]
+"""A :class:`str` (Unicode) based file or directory path."""
 
 TBD = Any
+"""Alias of :class:`~typing.Any`, when a type hint is meant to become more specific."""
+
 _T = TypeVar("_T")
+"""Type variable used internally in GitPython."""
+
+AnyGitObject = Union["Commit", "Tree", "TagObject", "Blob"]
+"""Union of the :class:`~git.objects.base.Object`-based types that represent actual git
+object types.
+
+As noted in :class:`~git.objects.base.Object`, which has further details, these are:
+
+* :class:`Blob <git.objects.blob.Blob>`
+* :class:`Tree <git.objects.tree.Tree>`
+* :class:`Commit <git.objects.commit.Commit>`
+* :class:`TagObject <git.objects.tag.TagObject>`
+
+Those GitPython classes represent the four git object types, per gitglossary(7):
+
+* "blob": https://git-scm.com/docs/gitglossary#def_blob_object
+* "tree object": https://git-scm.com/docs/gitglossary#def_tree_object
+* "commit object": https://git-scm.com/docs/gitglossary#def_commit_object
+* "tag object": https://git-scm.com/docs/gitglossary#def_tag_object
+
+For more general information on git objects and their types as git understands them:
+
+* "object": https://git-scm.com/docs/gitglossary#def_object
+* "object type": https://git-scm.com/docs/gitglossary#def_object_type
+
+:note:
+    See also the :class:`Tree_ish` and :class:`Commit_ish` unions.
+"""
 
 Tree_ish = Union["Commit", "Tree"]
-Commit_ish = Union["Commit", "TagObject", "Blob", "Tree"]
-Lit_commit_ish = Literal["commit", "tag", "blob", "tree"]
+"""Union of :class:`~git.objects.base.Object`-based types that are inherently tree-ish.
+
+See gitglossary(7) on "tree-ish": https://git-scm.com/docs/gitglossary#def_tree-ish
+
+:note:
+    This union comprises **only** the :class:`~git.objects.commit.Commit` and
+    :class:`~git.objects.tree.Tree` classes, **all** of whose instances are tree-ish.
+    This has been done because of the way GitPython uses it as a static type annotation.
+
+    :class:`~git.objects.tag.TagObject`, some but not all of whose instances are
```

Comment:
Thanks for mentioning this. It's great that this type includes what GitPython can handle though, as one day it might mean it can be taught to resolve tags as well.

Reply:
I had avoided changing this since it is not inherently a wrong definition of `Tree_ish`, but actually the situation you are describing, where it can be broader, is already the case. Furthermore, both `Commit_ish` and `Tree_ish` in `git.types` make a lot more sense if they both include `TagObject` rather than one of them including it and the other not. I've opened #1878 to add it.